toolkit/components/filewatcher/NativeFileWatcherMac.h

```diff
@@ -0,0 +1,255 @@
+ */
+struct CallBackAction{
+    CallBackAction(char*eventPath, const FSEventStreamEventFlags eventFlag, const FSEventStreamEventId eventIds)
+        : mEventFlags(eventFlag)
+        , mEventIds(eventIds)
+    {
+        snprintf(mEventPath, PATH_MAX, "%s", eventPath);
+    }
+
+    char mEventPath[PATH_MAX];
+    const FSEventStreamEventFlags mEventFlags;
+    const FSEventStreamEventId mEventIds;
+};
+
+
+/**
+ * The CallBackEvents struct manages the FSEvents events returned to the stream.
+ *
+ * This is populated inside of the FSEvents callback. It is then processed on a separate
+ * thread by NativeFileWatcherIOTask::RunInternal() which parses the events and calls the
+ * appropriate callbacks.
+ */
+struct CallBackEvents{
+    CallBackEvents()
+        : callBackLock("filewatcher::eventLock")
+    {}
+    std::queue<CallBackAction> mSavedEvents; // Used to hold individual FSEvents
+    Mutex callBackLock; // Used for safe access across multiple threads.
+};
+
+class NativeFileWatcherIOTask; // Forward Declaration of class.
+
+
+/**
+ * The NativeFileWatcherFSETask class which holds all of the functionality of the FSEventStream.
+ *
+ * It inherits from Runnable as we need to make a blocking call to CFRunLoopRun(). CFRunLoopRun() allows the
+ * stream to report file system events. The event stream is an immutable object, therefore with every change in the
+ * paths watched, the stream must be stopped and a new stream must be created with a new list of paths to be watched.
+ */
+class NativeFileWatcherFSETask : public Runnable
+{
+public:
+    explicit NativeFileWatcherFSETask(NativeFileWatcherIOTask* parent, CallBackEvents* cbe, std::vector<CFStringRef>& dirs);
+
+    // This is a blocking method which runs the steam on a CFRunLoop.
+    NS_IMETHOD Run() override;
+
+    // Adds path to the stream (and restarts it)
+    NS_IMETHOD AddPath(char* pathToAdd);
+
+    // Removes path from the stream and restarts it. Unless it removes the last path being watched by the stream.
+    // In this instance, we will return from this runnable without starting a new stream.
+    NS_IMETHOD RemovePath(char* pathToRemove);
+
+private:
+    //mozilla::Mutex runLoopLock;
+
+    FSEventStreamRef mEventStreamRef; // Reference to the FSEventStream
+    CFRunLoopRef mRunLoop = nullptr; // Reference to the RunLoop which is passed back to the parent class, to stop the RunLoop
+    std::vector<CFStringRef> mDirs; // A mutable list of directories which we build dynamically based on the needs of the stream
+    static CallBackEvents* cbe_internal; // Pointer to a callback events struct which is allocated in the parent class and used
+                                         // to hold events read from the stream for processing back in the parent class.
+    NativeFileWatcherIOTask* mParent; // Pointer to the parent class to allow stopping the CFRunLoop from another thread.

```

Comment:
Ok, so it's not a parent class. Maybe an owner?

What keeps each of these values alive?

Reply:
me to review

Reply:
This was a pretty gross/hack-y thing we had to do to get this working, and I don't like it as a long term solution. We have a blocking call which runs the Core Foundation run loop (see NativeFileWatcherFSETask::Run), which reports events from the filesystem. 

This means whatever thread the run loop is on (mFSEThread) can't stop the run loop itself through Mozilla's dispatch to worker thread framework, so it needs to be done from the scheduling thread (mIOThread). That means getting a reference for the run loop from the child class to it's owner.

As an alternative solution - I believe there's a version of CFRunLoopRun (the blocking call) which returns after a predetermined amount of time. The child class could then check for a flag signaling stop and choose to either reschedule its run loop or shut down. But that was learned in hindsight, and there won't be time to implement that.

Reply:
So the problem is for shutdown or for adding/removing files, right?

The technique I know of (tested with inotify, should also work with macOS) is to have an anonymous, in-memory, `control` pipe used for the sole purpose of waking up the loop. When you wish to stop the wait, write one byte to the `control` pipe. The blocking call will now return, the loop can detect that it is the `control` pipe, read a variable with the instructions, etc.